### PR TITLE
add ns deletion controller

### PR DESCRIPTION
**Pull request by qiujian16**
Signed-off-by: Jian Qiu <jqiu@redhat.com>

### CODE REVIEW

File: pkg/controller/clusternamespacedeletion/controller.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+	client   client.Client
+	recorder events.Recorder
+}
+
+// blank assignment to verify that ReconcileManagedCluster implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileClusterNamespaceDeletion{}
+
+func (r *ReconcileClusterNamespaceDeletion) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Name", request.Name)
+	reqLogger.Info("Reconciling the managed cluster meta object")
```
Comment (skeeey) [inline pkg/controller/clusternamespacedeletion/controller.go:53]: Reconciling the managed cluster namespace deletion?

Comment (skeeey) [inline pkg/controller/clusternamespacedeletion/controller.go:53]: Reconciling the managed cluster namespace deletion?

File: pkg/controller/clusternamespacedeletion/controller.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	if err := r.client.List(ctx, pods, client.InNamespace(ns.Name)); err != nil {
+		return reconcile.Result{}, err
+	}
+	validPods := filterPods(pods.Items, ns.Name)
+	if len(validPods) > 0 {
+		klog.Warningf("ManagedClusterNamespaceRemained", "There are some pods remaining in namespace %s", ns.Name)
+		return reconcile.Result{RequeueAfter: deletionGracePeriod}, nil
+	}
+
+	err = r.client.Delete(ctx, ns)
```
Comment (skeeey) [inline pkg/controller/clusternamespacedeletion/controller.go:136]: may need an event/log here?

File: pkg/controller/clusternamespacedeletion/manager.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		&source.Kind{Type: &clusterv1.ManagedCluster{}},
+		handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Name: o.GetName(),
+					},
+				},
+			}
+		}),
```
Comment (skeeey) [inline pkg/controller/clusternamespacedeletion/manager.go:62]: do we need a predicate here? only delete/update?

File: pkg/controller/clusternamespacedeletion/manager.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	); err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &corev1.Namespace{}},
```
Comment (skeeey) [inline pkg/controller/clusternamespacedeletion/manager.go:68]: It seems we don't need to watch the ns,

File: pkg/controller/clusternamespacedeletion/manager.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+				},
+			}
+		}),
+	); err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &hivev1.ClusterDeployment{}},
+		handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
```
Comment (skeeey) [inline pkg/controller/clusternamespacedeletion/manager.go:84]: do we need a predicate here? only delete/update?

File: pkg/controller/managedcluster/managedcluster_controller.go [hunk 6]
```diff
@@ -120,7 +110,7 @@ func (r *ReconcileManagedCluster) Reconcile(ctx context.Context, request reconci
 	}
 
 	// managed cluster is deleting, remove its namespace
```
Comment (skeeey) [inline pkg/controller/managedcluster/managedcluster_controller.go:112]: the comments is not right

File: pkg/controller/clusternamespacedeletion/manager.go (hunk 4 above)
Comment (qiujian16) [reply pkg/controller/clusternamespacedeletion/manager.go:82]: I think we need to in case controller restarted

**Comment by skeeey** [comment]
It's better to have e2e test for `infraenv` and `curator`, I think we can add the `infraenv` part firstly

**Pull request merged**